libnavui-dropin/src/test/java/com/mapbox/navigation/dropin/component/marker/GeocodingComponentTest.kt

```diff
@@ -0,0 +1,80 @@
+package com.mapbox.navigation.dropin.component.marker
+
+import com.mapbox.api.geocoding.v5.models.CarmenFeature
+import com.mapbox.geojson.Point
+import com.mapbox.navigation.base.ExperimentalPreviewMapboxNavigationAPI
+import com.mapbox.navigation.core.MapboxNavigation
+import com.mapbox.navigation.dropin.component.destination.DestinationAction.DidReverseGeocode
+import com.mapbox.navigation.dropin.component.destination.DestinationState
+import com.mapbox.navigation.dropin.component.destination.DestinationViewModel
+import com.mapbox.navigation.dropin.model.Destination
+import com.mapbox.navigation.dropin.util.Geocoder
+import com.mapbox.navigation.testing.MainCoroutineRule
+import io.mockk.MockKAnnotations
+import io.mockk.coEvery
+import io.mockk.every
+import io.mockk.impl.annotations.MockK
+import io.mockk.mockk
+import io.mockk.mockkObject
+import io.mockk.unmockkObject
+import io.mockk.verify
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.flow.MutableStateFlow
+import org.junit.After
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+
+@OptIn(ExperimentalPreviewMapboxNavigationAPI::class, ExperimentalCoroutinesApi::class)
+internal class GeocodingComponentTest {
+
+    @get:Rule
+    var coroutineRule = MainCoroutineRule()
+
+    lateinit var sut: GeocodingComponent
+
+    @MockK
+    lateinit var mockGeocoder: Geocoder
+
+    @MockK
+    lateinit var mockNavigation: MapboxNavigation
+
+    @MockK
+    lateinit var mockDestinationViewModel: DestinationViewModel
+    private lateinit var destinationStateFlow: MutableStateFlow<DestinationState>
+
+    @Before
+    fun setUp() {
+        MockKAnnotations.init(this, relaxUnitFun = true)
+        mockkObject(Geocoder)
+        every { Geocoder.create(any()) } returns mockGeocoder
+        every { mockNavigation.navigationOptions } returns mockk {
+            every { accessToken } returns "ACCESS_TOKEN"
+        }
+        destinationStateFlow = MutableStateFlow(DestinationState())
+        every { mockDestinationViewModel.state } returns destinationStateFlow
+
+        sut = GeocodingComponent(mockDestinationViewModel)
```

Comment:
There was a navigation-team discussion a few weeks ago. We essentially agreed to make "factory functions" for tests.

Essentially the goal is, to make it so each unit test can work in isolation. For example, what if you want to make a test where the initial DestinationState is different? The setup already assumes the initial state will be `DestinationState()`

Reply:
This is how test should be written. If you think you can improve this go ahead and change it.

Reply:
> This is how test should be written.

What do you mean exactly? I'm also commenting here because of the `LocationComponentTest` that [does not mockk the dependencies](https://github.com/mapbox/mapbox-navigation-android/pull/5576/files#diff-ac4e015a4964c380a6b15b8abbefd011c130011f27918b62d61d5e3f323f38bfR46). Making unit tests work in isolation is a common goal with unit tests, no?

The LocationComponentTest also sets up the map to [always load successful](https://github.com/mapbox/mapbox-navigation-android/pull/5576/files#diff-ac4e015a4964c380a6b15b8abbefd011c130011f27918b62d61d5e3f323f38bfR52). What if we want to make a test for when the map does not load successful? We will need to change that setup function.

If our team agrees on what makes good unit tests, we won't need to follow each others commits :).

Reply:
>Essentially the goal is, to make it so each unit test can work in isolation.

Junit always [runs tests](https://www.baeldung.com/junit-before-beforeclass-beforeeach-beforeall) as following to ensure test isolation:

```
@Before methods
first @Test method
@After methods
 
@Before methods
second @Test method
@After methods
```
Placing your common initialization code in `@Before` is a standard unit-testing practice.

> For example, what if you want to make a test where the initial DestinationState is different? The setup already assumes the initial state will be DestinationState()

In this particular example, you can either change initial state by setting `destinationStateFlow.value` before calling `sut.onAttached()` or by instantiating SUT with a different dependency in the test itself.

Reply:
But what about the `LocationComponentTest` instantiating a `NavigationLocationProvider`? That is not testing isolation, that's an instrumentation test.

These are nit picks, not a big deal for this test specifically - so feel free to merge. I'm referencing team discussions because, adding assumptions to the `@Before methods` eventually becomes an issue. The discussion was specific about these tests. Adding new tests that require different setup, creates the issue. https://github.com/mapbox/mapbox-navigation-android/blob/656204930b17bc437e57b4d8bedefd17bda74816/libnavigation-core/src/test/java/com/mapbox/navigation/core/trip/session/MapboxTripSessionTest.kt#L130-L171.

Reply:
> But what about the LocationComponentTest instantiating a NavigationLocationProvider?

Take a look at tested code.
```kotlin
// LocationComponent
override fun onAttached(mapboxNavigation: MapboxNavigation) {
        super.onAttached(mapboxNavigation)

        coroutineScope.launch {
            locationViewModel.firstLocation()
            mapView.getMapboxMap().getStyle {
                mapView.location.apply {
                    setLocationProvider(locationViewModel.navigationLocationProvider)
                    locationPuck = LocationPuck2D(
                        bearingImage = ContextCompat.getDrawable(
                            mapView.context,
                            R.drawable.mapbox_navigation_puck_icon
                        )
                    )
                    enabled = true
                }
            }
        }
    }
```
and its test

```kotlin
// LocationComponentTest
@Test
    fun `onAttach should configure and enable location component`() =
        coroutineRule.runBlockingTest {
            sut.onAttached(mockk())

            verify { mockLocationPlugin.setLocationProvider(locationProvider) }
            verify { mockLocationPlugin.enabled = true }
        }
```

Notice what we are verifying here is that the specific instance of the locationProvider `viewModel.locationViewModel.navigationLocationProvider` has been assigned to the location plugin (`setLocationProvider(locationViewModel.navigationLocationProvider)`).
It doesn't matter if we use real LocationProvider instance or a mock for this purpose. What matters is that it comes from locationViewModel.